### PR TITLE
Wrapper for OAuth URLs

### PR DESCRIPTION
**Pull request by bhavyay**
### - Link added should be a valid URL and accessible throughout the org.
   ### - If the branch name contains hotfix / revert by default the BVT workflow check will pass.

| Test Case Document URL                        |
|-----------------------------------------------|
| https://docs.google.com/spreadsheets/d/1VlHhQWwJBN0y3WMumA5upH2h52WYi61SxoTKr4XyXy8/edit?usp=sharing |

**PR Description:** Adding support for OAuth URLS(fetch authorize URL, access token, refresh token & revoke token APIs) in SDK. Also, adding support to access Razorpay Public APIs with access token
**Solution doc**: https://docs.google.com/document/d/1mW1Ql08vndWIB_PjS-HWM_R2Nr63isW3jPsewYbTycE/edit

### CODE REVIEW

File: documents/oauth_token.md [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+|----------------------|--------|---------------------------------------------------------------------------------------------------------------------------------------------------------|
+| client_id*           | string | Unique client identifier.                                                                                                                               |
+| redirect_uri*        | string | Callback URL used by Razorpay to redirect after the user approves or denies the authorisation request. The client should whitelist the 'redirect_uri'.  |
+| scopes*              | array  | Defines what access your application is requesting from the user. You can request one or multiple scopes by adding them to an array as indicated above. |
+| state*               | string | A random string generated by your service. This parameter helps prevent cross-site request forgery (CSRF) attacks.                                      |
+| onboarding_signature | string | A cryptographic string generated by your service using generateOnboardingSignature method in Utils class. Only applicable for accounts created with pre-fill KYC |
```
Comment (Swati707) [inline documents/oauth_token.md:32]: Does this formatting needs to be improved?
This is how it looks currently.
Screenshot 2024-02-28 at 11 48 50 AM

File: documents/oauth_token.md [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+    'client_secret'     => '<YOUR_CLIENT_SECRET>',
+    'redirect_uri'      => 'https://example.com/razorpay_callback',
+    'grant_type'        => 'authorization_code',
+    'code'              => '<AUTHORIZATION_CODE>',
+    'mode'              => 'test'
+}
+oauth_token = Razorpay::OAuthToken.get_access_token(options)
+```
+
+**Parameters:**
```
Comment (Swati707) [inline documents/oauth_token.md:53]: Same with this format.

File: documents/oauth_token.md [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+
+**Response:**
+```json
+{
+  "public_token": "rzp_test_oauth_9xu1rkZqoXlClS",
+  "token_type": "Bearer",
+  "expires_in": 7862400,
+  "access_token": "eyJ0eXAiOiJKV1QiLCJhbGciOiJSUzI1NiIsImp0aSI6IkY1Z0NQYkhhRzRjcUpnIn0.eyJhdWQiOiJGNFNNeEgxanMxbkpPZiIsImp0aSI6IkY1Z0NQYkhhRzRjcUpnIiwiaWF0IjoxNTkyODMxMDExLCJuYmYiOjE1OTI4MzEwMTEsInN1YiI6IiIsImV4cCI6MTYwMDc3OTgxMSwidXNlcl9pZCI6IkYycVBpejJEdzRPRVFwIiwibWVyY2hhbnRfaWQiOiJGMnFQaVZ3N0lNV01GSyIsInNjb3BlcyI6WyJyZWFkX29ubHkiXX0.Wwqt5czhoWpVzP5_aoiymKXoGj-ydo-4A_X2jf_7rrSvk4pXdqzbA5BMrHxPdPbeFQWV6vsnsgbf99Q3g-W4kalHyH67LfAzc3qnJ-mkYDkFY93tkeG-MCco6GJW-Jm8xhaV9EPUak7z9J9jcdluu9rNXYMtd5qxD8auyRYhEgs",
+  "refresh_token": "def50200f42e07aded65a323f6c53181d802cc797b62cc5e78dd8038d6dff253e5877da9ad32f463a4da0ad895e3de298cbce40e162202170e763754122a6cb97910a1f58e2378ee3492dc295e1525009cccc45635308cce8575bdf373606c453ebb5eb2bec062ca197ac23810cf9d6cf31fbb9fcf5b7d4de9bf524c89a4aa90599b0151c9e4e2fa08acb6d2fe17f30a6cfecdfd671f090787e821f844e5d36f5eacb7dfb33d91e83b18216ad0ebeba2bef7721e10d436c3984daafd8654ed881c581d6be0bdc9ebfaee0dc5f9374d7184d60aae5aa85385690220690e21bc93209fb8a8cc25a6abf1108d8277f7c3d38217b47744d7",
```
Comment (Swati707) [inline documents/oauth_token.md:70]:
```suggestion
  "access_token": "eyJ0e...",
  "refresh_token": "def50...",
```

File: lib/razorpay/validation_config.rb [hunk 4]
```diff
@@ -0,0 +1,19 @@
+module Razorpay
+  ValidationType = {
```
Comment (Swati707) [inline lib/razorpay/validation_config.rb:2]: Where is this getting used?

**Review by Swati707 — CHANGES_REQUESTED** [review]
Can we add proper documentations for all newly added methods? Atleast the public methods should be documented.

File: lib/razorpay/validation_config.rb (hunk 4 above)
Comment (bhavyay) [reply lib/razorpay/validation_config.rb:2]: https://github.com/razorpay/razorpay-ruby/blob/be0455b3719eab23913e6fca28512c869782199b/lib/razorpay/payload_validator.rb#L28

File: documents/oauth_token.md (hunk 1 above)
Comment (bhavyay) [reply documents/oauth_token.md:32]: Formatting was done. Had to add additional line after parameters which was causing this issue.

File: documents/oauth_token.md (hunk 2 above)
Comment (bhavyay) [reply documents/oauth_token.md:53]: Formatting was done. Had to add additional line after parameters which was causing this issue.

File: documents/oauth_token.md (hunk 3 above)
Comment (bhavyay) [reply documents/oauth_token.md:70]: The response was taken from razorpay public docs. This might easier for partners to understand

**Review by Swati707 — APPROVED** [review]
Discussed the documentation comment on call. There is no such method and class level code documentation added for existing files in this repo. And these methods are direct calls to Rzp public APIs. Hence won’t be adding for the new methods. 

Rest LGTM.

**Pull request merged**